### PR TITLE
[NTOS:SE] Implement security access checks by object type list

### PR DESCRIPTION
**Pull request by GeoB99**
Windows offers a mechanism to protect resource objects through what is called **access check by type**.  This kind of access check uses object type ACEs to protect objects belonging to a directory service. This might as well allow the possibility for the implementation of AD (Active Directory). 

ReactOS already provides the code that creates such ACEs but no access check mechanism is written for that, which this is the aim of this PR is  trying to achieve for.
## Proposed Changes
- Update `NtAccessCheck`, `NtAccessCheckByType` and `NtAccessCheckByTypeResultList` syscalls SAL2 annotation. Inspired from Process Hacker.
- Implement the object type list core utilities (`objtype.c`).
- `SepAccessCheck` will now be `SepAccessCheckWorker`, a worker helper function that further abstracts the access check mechanism in the kernel. Whereas the `SepAccessCheck` name will be used as a centralized function used by the access check NT system calls, thus avoiding boilerplate code.
- Fix a kernel crash when capturing data from a security descriptor, namely the owner and group. `SepGetSDOwner` and `SepGetSDGroup` were buggy as they would pass to the caller garbage data due to the way it actually returns such data if either the group or owner is NULL. We already have `SepGetOwnerFromDescriptor` and `SepGetGroupFromDescriptor` which are defined in the private `se.h` kernel header. Plus they are way better, so deprecate the former functions.
## TODO
- [x] Finish writing tests for `NtAccessCheckByType` and begin writing tests for `NtAccessCheckByTypeResultList`

### CODE REVIEW

File: ntoskrnl/se/objtype.c [hunk 1]
```diff
@@ -99,15 +372,14 @@ SeCaptureObjectTypeList(
...
 SeReleaseObjectTypeList(
-    _In_  _Post_invalid_ POBJECT_TYPE_LIST CapturedObjectTypeList,
+    _In_  _Post_invalid_ POBJECT_TYPE_LIST_INTERNAL CapturedObjectTypeList,
     _In_ KPROCESSOR_MODE PreviousMode)
 {
+    PAGED_CODE();
+
     if ((PreviousMode != KernelMode) && (CapturedObjectTypeList != NULL))
         ExFreePoolWithTag(CapturedObjectTypeList, TAG_SEPA);
 }
```
Comment (binarymaster) [inline ntoskrnl/se/objtype.c:385]:
```
../src/ntoskrnl/se/objtype.c:385:1: error: expected declaration or statement at end of input
 }
 ^
In file included from ../src/sdk/lib/pseh/include/pseh/pseh2.h:90,
                 from ../src/ntoskrnl/include/ntoskrnl.h:82,
                 from ../src/ntoskrnl/se/objtype.c:11:
```

**Review by binarymaster — COMMENTED** [review]
Something GCC...

File: modules/rostests/apitests/ntdll/NtAccessCheckByType.c [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+ * PURPOSE:         Tests for the NtAccessCheckByType API
+ * COPYRIGHT:       Copyright 2023 George Bișoc <george.bisoc@reactos.org>
+ */
+
+#include "precomp.h"
+
+GENERIC_MAPPING RegMapping = {KEY_READ, KEY_WRITE, KEY_EXECUTE, KEY_ALL_ACCESS};
+GUID ObjectType = {0x12345678, 0x1234, 0x5678, {0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88}};
+GUID ChildObjectType = {0x23456789, 0x2345, 0x6786, {0x2, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99}};
+static SID_IDENTIFIER_AUTHORITY WorldAuthority = {SECURITY_WORLD_SID_AUTHORITY};
```
Comment (binarymaster) [inline modules/rostests/apitests/ntdll/NtAccessCheckByType.c:13]:
```
../src/modules/rostests/apitests/ntdll/NtAccessCheckByType.c:13:33: error: 'WorldAuthority' defined but not used [-Werror=unused-variable]
 static SID_IDENTIFIER_AUTHORITY WorldAuthority = {SECURITY_WORLD_SID_AUTHORITY};
                                 ^~~~~~~~~~~~~~
```

**Comment by GeoB99** [comment]
Ready.

**ReactOS**
Capture
**Windows Server 2003 SP2**
Capture2
**Windows 10 (CMD ran as admin)**
Capture3

**Comment by GeoB99** [comment]
Test KVM: https://build.reactos.org/#/builders/6/builds/33822
Test Vbox: https://build.reactos.org/#/builders/3/builds/21296

Test results: https://reactos.org/testman/compare.php?ids=88694,88699

**Comment by learn-more** [comment]
> Test KVM: https://build.reactos.org/#/builders/6/builds/33822 Test Vbox: https://build.reactos.org/#/builders/3/builds/21296
> 
> Test results: https://reactos.org/testman/compare.php?ids=88694,88699

The 2nd test has canceled / crashed, is this caused by your changes?

**Comment by GeoB99** [comment]
> > Test KVM: https://build.reactos.org/#/builders/6/builds/33822 Test Vbox: https://build.reactos.org/#/builders/3/builds/21296
> > Test results: https://reactos.org/testman/compare.php?ids=88694,88699
> 
> The 2nd test has canceled / crashed, is this caused by your changes?

No.

**Pull request merged**